contrib/scalafix/src/python/pants/contrib/scalafix/tasks/scalafix.py

```
@@ -0,0 +1,123 @@
+                            JarDependency(org='ch.epfl.scala', name='scalafix-cli_2.11.11', rev='0.4.2'),
+                          ])
+
+  @classmethod
+  def prepare(cls, options, round_manager):
+    super(ScalaFix, cls).prepare(options, round_manager)
+    round_manager.require_data('runtime_classpath')
+
+  @abstractmethod
+  def _finalize_target(self, target, results_dir):
+    """Given a Target and a successfully populated results_dir containing its sources, finalize."""
+
+  def _is_fixed(self, target):
+    return target.has_sources(self._SCALA_SOURCE_EXTENSION) and (not target.is_synthetic)
+
+  def execute(self):
+    classpaths = self.context.products.get_data('runtime_classpath')
+    # NB: This task uses only the literal classpath of each target, so does not need
+    # `invalidate_dependents=True`.
+    with self.invalidated(self.context.targets(self._is_fixed)) as invalidation_check:
+      for vt in invalidation_check.all_vts:
+        if not vt.valid:
+          self.context.log.debug('Fixing {}...'.format(vt.target.address.spec))
+          classpath = [e for _, e in classpaths.get_for_target(vt.target)]
+          self._run(vt.target, vt.results_dir, classpath)
+        else:
+          self.context.log.debug('Already fixed {}'.format(vt.target.address.spec))
+        # Target outputs are valid: finalize them.
+        self._finalize_target(vt.target, vt.results_dir)
+
+  def _run(self, target, results_dir, classpath):
+    # We always operate on copies of the files, so we execute in place.
+    args = ['--in-place']
+    args.append('--sourceroot={}'.format(results_dir))
+    args.append('--classpath={}'.format(':'.join(classpath)))
+    if self.get_options().config:
+      args.append('--config={}'.format(self.get_options().config))
+    if self.get_options().rewrites:
+      args.append('--rewrites={}'.format(self.get_options().rewrites))
+    if self.get_options().level == 'debug':
+      args.append('--verbose')
```

scalafix will still in some cases silently fail on misconfigured semanticdbs, https://github.com/scalacenter/scalafix/issues/264 This is a bug, scalafix should report helpful error messages when things go wrong and report the appropriate exit code to accompany it. For example, scalafix will exit with code 32 when it encounters a "stale semanticdb" (file contents doesn't match semanticdb contents).